pkg/render/manager_route_config.go

```diff
@@ -0,0 +1,416 @@
+	// that will be mounted by the VoltronRouteConfig so Voltron can be restarted if the value changes.
+	AddConfigMap(configMap *corev1.ConfigMap)
+
+	// AddSecret accepts a Secret referenced by a TLS terminated route. This is used to detect changes to Secrets
+	// that will be mounted by the VoltronRouteConfig so Voltron can be restarted if the value changes.
+	AddSecret(secret *corev1.Secret)
+	Build() (*VoltronRouteConfig, error)
+}
+
+type voltronRouteConfigBuilder struct {
+	mountedConfigMaps map[string]struct{}
+	mountedSecrets    map[string]struct{}
+
+	volumeMounts    []corev1.VolumeMount
+	volumes         []corev1.Volume
+	hashableContent []interface{}
+
+	configMaps map[string]*corev1.ConfigMap
+	secrets    map[string]*corev1.Secret
+
+	tlsTerminatedRoutes  []operatorv1.TLSTerminatedRoute
+	tlsPassThroughRoutes []operatorv1.TLSPassThroughRoute
+}
+
+func NewVoltronRouteConfigBuilder() VoltronRouteConfigBuilder {
+	return &voltronRouteConfigBuilder{
+		configMaps:        map[string]*corev1.ConfigMap{},
+		secrets:           map[string]*corev1.Secret{},
+		mountedConfigMaps: map[string]struct{}{},
+		mountedSecrets:    map[string]struct{}{},
+	}
+}
+
+func (builder *voltronRouteConfigBuilder) AddConfigMap(configMap *corev1.ConfigMap) {
+	if _, ok := builder.configMaps[configMap.Name]; !ok {
+		builder.configMaps[configMap.Name] = configMap
+	}
+}
+
+func (builder *voltronRouteConfigBuilder) AddSecret(secret *corev1.Secret) {
+	if _, ok := builder.secrets[secret.Name]; !ok {
+		builder.secrets[secret.Name] = secret
+	}
+}
+
+func (builder *voltronRouteConfigBuilder) AddTLSTerminatedRoute(route operatorv1.TLSTerminatedRoute) {
+	builder.tlsTerminatedRoutes = append(builder.tlsTerminatedRoutes, route)
+}
+
+func (builder *voltronRouteConfigBuilder) AddTLSPassThroughRoute(route operatorv1.TLSPassThroughRoute) {
+	builder.tlsPassThroughRoutes = append(builder.tlsPassThroughRoutes, route)
+}
+
+func (builder *voltronRouteConfigBuilder) Build() (*VoltronRouteConfig, error) {
+	var uiTLSTerminatedRoutes []*tlsTerminatedRoute
+	var tunnelTLSTerminatedRoutes []*tlsTerminatedRoute
+	var tunnelTLSPassThroughRoutes []*tlsPassThroughRoute
+
+	for _, route := range builder.tlsTerminatedRoutes {
+		if route.Spec.CABundle == nil {
+			return nil, fmt.Errorf("CABundle is required")
+		}
+
+		r := &tlsTerminatedRoute{
+			Destination: route.Spec.Destination,
+			Path:        route.Spec.PathMatch.Path,
+			PathRegexp:  route.Spec.PathMatch.PathRegexp,
+			PathReplace: route.Spec.PathMatch.PathReplace,
+		}
+
+		if route.Spec.Target == operatorv1.TargetTypeUI {
+			uiTLSTerminatedRoutes = append(uiTLSTerminatedRoutes, r)
+			if route.Spec.Unauthenticated != nil {
+				r.Unauthenticated = *route.Spec.Unauthenticated
+			}
+		} else if route.Spec.Target == operatorv1.TargetTypeUpstreamTunnel {
+			tunnelTLSTerminatedRoutes = append(tunnelTLSTerminatedRoutes, r)
+		} else {
+			return nil, fmt.Errorf("unknown Target value %s", route.Spec.Target)
```

Comment:
Seems like validation is spread throughout the implementation of the render code for this particular API. 

Not necessarily against it.. but it is a departure from the existing model where validation is performed ASAP. Was this a conscious decision? I am finding it makes it harder for me as the reviewer to make sure that all the important pieces of validation are enforced, since I don't know exactly where to look.

Reply:
All the validation here is performed in the Build function (well with the exception of retrieving the secrets / config maps, but I could move it to here), and I think this is common with Builder patterns.

The initial implementation did validation and parsing out the volumes / volume mounts as the Route objects were passed in, but it spread out some of the validation between when you were adding the routes and when you were building the final object.

It might have just been a complexity in a previous iteration, but I like that all the validation and parsing is contained to the single Build() function.